File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.py
def solution(k, ranges):
    answer = []
    hail_arr = [k]
    # 우박수열 만들기
    while k>1:
        if k%2==0:
            k//=2
        else:
            k=k*3+1
        hail_arr.append(k)
        
    for a, b in ranges:
        if b<=0: # [a,-b]인 경우
            b = len(hail_arr) -1 + b
        
        if a>b: # 유효하지 않은 구간일 경우
            answer.append(-1.0)
            continue
        
        h = max(hail_arr[a:b+1]) # 사각형 높이
        area = (b-a) * h
        for i in range(a,b):
            area -= 1 * abs(hail_arr[i]-hail_arr[i+1]) / 2 # 빈 공간 삼각형 제거
            area -= (h-max(hail_arr[i],hail_arr[i+1])) * 1 # 빈 공간 사각형 제거
        answer.append(area)
            
    return answer

File: 프로그래머스/2/134239. 우박수열 정적분/우박수열 정적분.js
function solution(k, ranges) {
    const result = []
    const [y, accr] = [[k], [0]]
    // 콜라츠 추측
    while (k > 1) {
        if (k % 2 === 0) k /= 2
        else k = 3 * k + 1
        y.push(k)
    }
    const last = y.length - 1 // 끝점 index
    // 넓이 구간합 계산후 저장
    for (let i = 1; i <= last; i += 1) {
        accr[i] = accr[i - 1] + (y[i - 1] + y[i]) / 2
    }
    // x2와 x1에서의 구간합의 차이(도형의 넓이) 넣기
    for (const range of ranges) {
        const [x1, x2] = [range[0], last + range[1]]
        if (x1 > x2) result.push(-1)
        else result.push(accr[x2] - accr[x1])
    }
    return result
}